# Project: Database

**Instructions for Students:**

Please carefully follow these steps to complete and submit your project:

1. **Completing the Project**: You are required to work on and complete all tasks in the provided project. Be disciplined and ensure that you thoroughly engage with each task.
   
2. **Creating a Google Drive Folder**: Each of you must create a new folder on your Google Drive if you haven't already. This will be the repository for all your completed assignment and project files, aiding you in keeping your work organized and accessible.
   
3. **Uploading Completed Project**: Upon completion of your project, make sure to upload all necessary files, involving codes, reports, and related documents into the created Google Drive folder. Save this link in the 'Student Identity' section and also provide it as the last parameter in the `submit` function that has been provided.
   
4. **Sharing Folder Link**: You're required to share the link to your project Google Drive folder. This is crucial for the submission and evaluation of your project.
   
5. **Setting Permission toPublic**: Please make sure your Google Drive folder is set to public. This allows your instructor to access your solutions and assess your work correctly.

Adhering to these procedures will facilitate a smooth project evaluation process for you and the reviewers.

## Project Description

In this project, you'll work with the Kaggle E-commerce dataset. Your tasks are to import this dataset into a SQLite database, explore it using SQL to identify key attributes and patterns, and perform basic data cleaning procedures. This project is designed to improve your skills in database management, SQL usage, and data cleaning.

Rmember to make a copy of this notebook in your Google Drive and work in your own copy.

## I. Loading and Querying with SQL

In this task, you will use Kaggle E-commerce dataset. You need to download it from the link here: [Kaggle E-commerce data](https://www.kaggle.com/datasets/carrie1/ecommerce-data?select=data.csv).

After you download the dataset, you will then import the dataset into SQL format, using SQLite.

### I.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [2]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below

import csv
import sqlite3
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
try:
    df_data = pd.read_csv('/content/drive/MyDrive/SKILL ACADEMY/PROJECT 7/data.csv', encoding='utf-8')
except UnicodeDecodeError:
    try:
        print('halo')
        df_data = pd.read_csv('/content/drive/MyDrive/SKILL ACADEMY/PROJECT 7/data.csv', encoding='latin1')
    except UnicodeDecodeError:
        try:
            df_data = pd.read_csv('/content/drive/MyDrive/SKILL ACADEMY/PROJECT 7/data.csv', encoding='ISO-8859-1')
        except UnicodeDecodeError:
            df_data = pd.read_csv('/content/drive/MyDrive/SKILL ACADEMY/PROJECT 7/data.csv', encoding='cp1252')

df_data.head()

halo


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [ ]:
print(df_data['CustomerID'].value_counts())

17841.0    7983
14911.0    5903
14096.0    5128
12748.0    4642
14606.0    2782
           ... 
15070.0       1
15753.0       1
17065.0       1
16881.0       1
16995.0       1
Name: CustomerID, Length: 4372, dtype: int64


### I.2. Data Loading and SQL CREATE and INSERT to database table

Now you can create the database and tables. Follow the same column naming as the dataset.

Next insert all the data from the dataset into the table in the database.

In [ ]:
# Write your data loading to SQL database here

conn = sqlite3.connect('ecommerce.db')

cursor = conn.cursor()

cursor.execute(
    '''
        CREATE TABLE IF NOT EXISTS ecommerce_data(
            ID            INTEGER PRIMARY KEY AUTOINCREMENT,
            InvoiceNo     TEXT,
            StockCode     TEXT,
            Description   TEXT,
            Quantity      INTEGER,
            InvoiceDate   TEXT, # pada sqlite tidak ada tipe data date
            UnitPrice     REAL,
            CustomerID    INTEGER,
            Country       TEXT
        )
    '''
)

conn.commit()

In [ ]:
pd.read_sql_query("SELECT * FROM ecommerce_data", conn)

Empty DataFrame
Columns: [ID, InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [ ]:
with open('/content/drive/MyDrive/SKILL ACADEMY/PROJECT 7/data.csv', 'r', encoding='latin1') as f:

    # Skip the header row
    next(f)

    cursor.executemany(
        '''
            INSERT INTO ecommerce_data (InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        ''', csv.reader(f))

    conn.commit()

In [ ]:
pd.read_sql_query("SELECT * FROM ecommerce_data", conn)

ID InvoiceNo StockCode                          Description  \
0            1    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER   
1            2    536365     71053                  WHITE METAL LANTERN   
2            3    536365    84406B       CREAM CUPID HEARTS COAT HANGER   
3            4    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE   
4            5    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.   
...        ...       ...       ...                                  ...   
541904  541905    581587     22613          PACK OF 20 SPACEBOY NAPKINS   
541905  541906    581587     22899         CHILDREN'S APRON DOLLY GIRL    
541906  541907    581587     23254        CHILDRENS CUTLERY DOLLY GIRL    
541907  541908    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE   
541908  541909    581587     22138        BAKING SET 9 PIECE RETROSPOT    

        Quantity      InvoiceDate  UnitPrice CustomerID         Country  
0              6   12/1/2010 8:26       2.55      17850  United Kingdom  
1              6   12/1/2010 8:26       3.39      17850  United Kingdom  
2              8   12/1/2010 8:26       2.75      17850  United Kingdom  
3              6   12/1/2010 8:26       3.39      17850  United Kingdom  
4              6   12/1/2010 8:26       3.39      17850  United Kingdom  
...          ...              ...        ...        ...             ...  
541904        12  12/9/2011 12:50       0.85      12680          France  
541905         6  12/9/2011 12:50       2.10      12680          France  
541906         4  12/9/2011 12:50       4.15      12680          France  
541907         4  12/9/2011 12:50       4.15      12680          France  
541908         3  12/9/2011 12:50       4.95      12680          France  

[541909 rows x 9 columns]

### I.3. Verifying the database

Before we go to the next section, let's check our database first. You need to change the db_name and table_name with your database name and table name, afterwards you can just run the code block. The result should be:

- Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
- Number of rows: 541909

In [ ]:
# import sqlite3

# # Connect to SQLite database
# conn = sqlite3.connect('db_name.sqlite') #change db_name to your database name.
# cur = conn.cursor()

# # Get column names
# cur.execute("PRAGMA table_info(table_name)") #change table_name to your table name
# sql_columns = [column[1] for column in cur.fetchall()] #the variable sql_columns is used for submission, do not change the variable name
# print(f"Columns: {sql_columns}")

# # Get total number of rows
# cur.execute("SELECT COUNT(*) FROM table_name") #change table_name to your table name
# sql_num_rows = cur.fetchone()[0] #the variable sql_num_rows is used for submission, do not change the variable name
# print(f"Number of rows: {sql_num_rows}")

# # Close the connection
# conn.close()

In [ ]:
# Get column names
cursor.execute(
    '''
        PRAGMA table_info(ecommerce_data)
    '''
)

sql_columns = [column[1] for column in cursor.fetchall()]
sql_columns.pop(0)
print(f"Columns: {sql_columns}")

Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


In [ ]:
cursor.execute(
    '''
        SELECT COUNT(*) FROM ecommerce_data
    '''
)

sql_num_rows = cursor.fetchone()[0]
print(f"Number of rows: {sql_num_rows}")

Number of rows: 541909


### I.3.4. Running SQL Queries

Once we have the database in SQL format, let's do some data exploration and analysis.

You are part of the engineering team at a global e-commerce company. Your company has a vast catalogue of products and serves customers around the globe. As the company continues to grow, the management decides it's time to expand the business even more.

As an essential part of this expansion, your team is tasked with diving into your company's rich e-commerce dataset to extract crucial insights that will guide the expansion strategy.

1. **Total number of unique products and unique customers:**

   Your first task is to gauge the breadth of your operations. By determining the total number of unique products you sell and the total number of unique customers you serve, you can assess the scale and diversity of your business.

2. **Total revenue for each product:**

   Next, you're going to identify the star performers in your product portfolio. Calculating the total revenue for each product will reveal which items are the biggest revenue drivers. These insights can help guide decisions about product focus and marketing efforts. Limit the result to the top 5 product with the biggest revenue.

3. **Most profitable countries (by total sales):**

   To pinpoint the most fruitful geographic areas for your business, you'll identify which countries generate the most profit. This data will help you understand where your operations are already strong and where there might be potential for regional expansion. Limit the result to the top 5 countries with the biggest profit.

4. **Top-selling products (by total sales) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by sales for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

4. **Top-selling products (by quantity) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by quantity for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

By answering these queries, your team will provide valuable data-driven insights that will directly contribute to the strategic decisions about where and how your business expands. As part of the engineering team, you're not just maintaining the technical infrastructure - you're shaping the future of the company.

> Hint: The Product name is stored in column Descrioption

In [ ]:
# Write your SQL Query here

#use the following variable name to store the result from your SQL Query
sql_num_products = 0
sql_num_customers = 0
sql_top_product_revenues = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_countries = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_selling_products_by_country_by_sales = {} #dictionary where key is country name, value is a list of top-selling products by sales, see example below
sql_top_selling_products_by_country_by_quantity = {} #dictionary where key is country name, value is a list of top-selling products by quantity, see example below


- https://www.w3schools.com/sql/sql_distinct.asp

In [ ]:
# 1. Find the total number of unique products and unique customers
# Example: sql_num_products = 1559
# Example: sql_num_customers = 1675


cursor.execute(
    '''
        SELECT COUNT(DISTINCT StockCode) AS total_unique_products FROM ecommerce_data
    '''
)

total_unique_products = cursor.fetchone()[0]
print(f"Total unique products: {total_unique_products}")

cursor.execute(
    '''
        SELECT COUNT(DISTINCT CustomerID) AS total_unique_customers FROM ecommerce_data
    '''
)

total_unique_customers = cursor.fetchone()[0] - 1
print(f"Jumlah customer unik: {total_unique_customers}")



sql_num_products = total_unique_products
sql_num_customers = total_unique_customers

Total unique products: 4070
Jumlah customer unik: 4372


- https://www.w3schools.com/sql/sql_sum.asp
- https://www.w3schools.com/sql/sql_orderby.asp

In [ ]:
# 2. Calculate the total revenue for each product, limited to top 5
# Example: sql_top_product_revenues = [('PRODUCT 1', 149385.12), ('PRODUCT 2', 54895.53), ('PRODUCT 3', 44545.55), ('PRODUCT 4', 38000.00), ('PRODUCT 5', 36000.00)]

cursor.execute(
    '''
        SELECT Description, SUM(Quantity * UnitPrice) AS total_revenue FROM ecommerce_data
        GROUP BY Description
        ORDER BY total_revenue DESC
        LIMIT 5

    '''
)

top_product_revenues = cursor.fetchall()

print("Top 5 produk dengan pendapatan terbesar:")

for row in top_product_revenues:
    stock_code, total_revenue = row
    print(f"Stock Code: {stock_code}, Total Revenue: {total_revenue}")

print(top_product_revenues)



sql_top_product_revenues = top_product_revenues

print(sql_top_product_revenues)


Top 5 produk dengan pendapatan terbesar:
Stock Code: DOTCOM POSTAGE, Total Revenue: 206245.48000000019
Stock Code: REGENCY CAKESTAND 3 TIER, Total Revenue: 164762.19
Stock Code: WHITE HANGING HEART T-LIGHT HOLDER, Total Revenue: 99668.46999999849
Stock Code: PARTY BUNTING, Total Revenue: 98302.98000000048
Stock Code: JUMBO BAG RED RETROSPOT, Total Revenue: 92356.03000000204
[('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 164762.19), ('WHITE HANGING HEART T-LIGHT HOLDER', 99668.46999999849), ('PARTY BUNTING', 98302.98000000048), ('JUMBO BAG RED RETROSPOT', 92356.03000000204)]
[('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 164762.19), ('WHITE HANGING HEART T-LIGHT HOLDER', 99668.46999999849), ('PARTY BUNTING', 98302.98000000048), ('JUMBO BAG RED RETROSPOT', 92356.03000000204)]


In [ ]:
# 3. Identify the most profitable countries (by total sales), limited to top 5
# Example: sql_top_countries = [('COUNTRY 1', 600000.00), ('COUNTRY 2', 334857.45), ('COUNTRY 3', 245879.00), ('COUNTRY 4', 180000.00), ('COUNTRY 5', 164389.45)]

cursor.execute(
    '''
       SELECT Country, SUM(Quantity * UnitPrice) AS total_revenue FROM ecommerce_data
       GROUP BY Country
       ORDER BY total_revenue DESC
       LIMIT 5
    '''
)

top_country_profit = cursor.fetchall()

print("Top 5 negara dengan profit terbesar")

for row in top_country_profit:
    country, total_profit = row
    print(f"Country: {country}, Total Profit: {total_profit}")

print(top_country_profit)



sql_top_countries = top_country_profit

print(sql_top_countries)


Top 5 negara dengan profit terbesar
Country: United Kingdom, Total Profit: 8187806.364001113
Country: Netherlands, Total Profit: 284661.54000000015
Country: EIRE, Total Profit: 263276.81999999826
Country: Germany, Total Profit: 221698.20999999862
Country: France, Total Profit: 197403.90000000002
[('United Kingdom', 8187806.364001113), ('Netherlands', 284661.54000000015), ('EIRE', 263276.81999999826), ('Germany', 221698.20999999862), ('France', 197403.90000000002)]
[('United Kingdom', 8187806.364001113), ('Netherlands', 284661.54000000015), ('EIRE', 263276.81999999826), ('Germany', 221698.20999999862), ('France', 197403.90000000002)]


- https://www.digitalocean.com/community/tutorials/how-to-use-the-sqlite3-module-in-python-3

In [ ]:
# 4. Find the top-selling products (by total sales) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_sales = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 200000.00), ('COUNTRY 1 PRODUCT 2', 100000.00), ('COUNTRY 1 PRODUCT 3', 95348.00), ('COUNTRY 1 PRODUCT 4', 90000.43), ('COUNTRY 1 PRODUCT 5', 80000.00)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000.00), ('COUNTRY 2 PRODUCT 2', 7345.00), ('COUNTRY 2 PRODUCT 3', 6934.23), ('COUNTRY 2 PRODUCT 4', 6800.23), ('COUNTRY 2 PRODUCT 5', 4000.00)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 7442.84), ('COUNTRY 5 PRODUCT 2', 5175.00), ('COUNTRY 5 PRODUCT 3', 3098.0), ('COUNTRY 5 PRODUCT 4', 3014.34), ('COUNTRY 5 PRODUCT 5', 2857.83)],


top_selling_products_by_country = {}

for country, _ in top_country_profit:

    cursor.execute(
        '''
            SELECT Description, SUM(Quantity * UnitPrice) AS total_revenue FROM ecommerce_data
            WHERE Country = ?
            GROUP BY Description
            ORDER BY total_revenue DESC
            LIMIT 5
        ''', (country,)
    )

    top_selling_products = cursor.fetchall()

    top_selling_products_by_country[country] = top_selling_products


for country, top_products in top_selling_products_by_country.items():
    print(f"Top 5 penjualan produk di {country}:")
    for stock_code, total_sales in top_products:
        print(f"Stock Code: {stock_code}, Total Penjualan: {total_sales}")
    print()

print(top_selling_products_by_country)



sql_top_selling_products_by_country_by_sales = top_selling_products_by_country

print(sql_top_selling_products_by_country_by_sales)

Top 5 penjualan produk di United Kingdom:
Stock Code: DOTCOM POSTAGE, Total Penjualan: 206245.48000000019
Stock Code: REGENCY CAKESTAND 3 TIER, Total Penjualan: 134405.93999999992
Stock Code: WHITE HANGING HEART T-LIGHT HOLDER, Total Penjualan: 93953.0699999987
Stock Code: PARTY BUNTING, Total Penjualan: 92501.73000000033
Stock Code: JUMBO BAG RED RETROSPOT, Total Penjualan: 84516.44000000166

Top 5 penjualan produk di Netherlands:
Stock Code: RABBIT NIGHT LIGHT, Total Penjualan: 9568.48
Stock Code: ROUND SNACK BOXES SET OF4 WOODLAND , Total Penjualan: 7991.400000000001
Stock Code: SPACEBOY LUNCH BOX , Total Penjualan: 7485.5999999999985
Stock Code: DOLLY GIRL LUNCH BOX, Total Penjualan: 6828.599999999999
Stock Code: ROUND SNACK BOXES SET OF 4 FRUITS , Total Penjualan: 4039.2000000000007

Top 5 penjualan produk di EIRE:
Stock Code: REGENCY CAKESTAND 3 TIER, Total Penjualan: 7442.849999999997
Stock Code: CARRIAGE, Total Penjualan: 5175.0
Stock Code: JAM MAKING SET WITH JARS, Total Penju

In [ ]:
# 5. Find the top-selling products (by quantity) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_quantity = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 22343), ('COUNTRY 1 PRODUCT 2', 12345), ('COUNTRY 1 PRODUCT 3', 9534), ('COUNTRY 1 PRODUCT 4', 9000), ('COUNTRY 1 PRODUCT 5', 8000)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000), ('COUNTRY 2 PRODUCT 2', 7345), ('COUNTRY 2 PRODUCT 3', 6934), ('COUNTRY 2 PRODUCT 4', 6800), ('COUNTRY 2 PRODUCT 5', 4000)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 48345), ('COUNTRY 5 PRODUCT 2', 43965), ('COUNTRY 5 PRODUCT 3', 43556), ('COUNTRY 5 PRODUCT 4', 34567), ('COUNTRY 5 PRODUCT 5', 33436)],


top_quantity_products_by_country= {}

for country, _ in top_country_profit:

    cursor.execute(
        '''
            SELECT Country, Description, SUM(Quantity) AS quantity FROM ecommerce_data
            WHERE Country = ?
            GROUP BY Description
            ORDER BY quantity DESC
            LIMIT 5
        ''', (country,)
    )

    top_quantity_products = cursor.fetchall()

    top_quantity_products_by_country[country] = top_quantity_products


# for country, top_products in top_quantity_products_by_country.items():
#     print(f"Top 5 penjualan produk di {country}:")
#     for product, quantity in top_products:
#         print(f"Produk: {product}, Total Penjualan: {quantity}")
#     print()

print(top_quantity_products_by_country)



sql_top_selling_products_by_country_by_quantity  = top_quantity_products_by_country

print(sql_top_selling_products_by_country_by_quantity )

{'United Kingdom': [('United Kingdom', 'WORLD WAR 2 GLIDERS ASSTD DESIGNS', 48326), ('United Kingdom', 'JUMBO BAG RED RETROSPOT', 43167), ('United Kingdom', 'POPCORN HOLDER', 34365), ('United Kingdom', 'ASSORTED COLOUR BIRD ORNAMENT', 33679), ('United Kingdom', 'WHITE HANGING HEART T-LIGHT HOLDER', 33193)], 'Netherlands': [('Netherlands', 'RABBIT NIGHT LIGHT', 4801), ('Netherlands', 'SPACEBOY LUNCH BOX ', 4528), ('Netherlands', 'DOLLY GIRL LUNCH BOX', 4132), ('Netherlands', 'PACK OF 72 RETROSPOT CAKE CASES', 4128), ('Netherlands', 'ROUND SNACK BOXES SET OF4 WOODLAND ', 3132)], 'EIRE': [('EIRE', 'PACK OF 72 RETROSPOT CAKE CASES', 1728), ('EIRE', '60 TEATIME FAIRY CAKE CASES', 1536), ('EIRE', 'VINTAGE SNAP CARDS', 1492), ('EIRE', 'ASSORTED INCENSE PACK', 1440), ('EIRE', 'PACK OF 60 PINK PAISLEY CAKE CASES', 1296)], 'Germany': [('Germany', 'ROUND SNACK BOXES SET OF4 WOODLAND ', 1218), ('Germany', 'ASSORTED COLOURS SILK FAN', 1164), ('Germany', 'POSTAGE', 1104), ('Germany', 'WOODLAND CHARL

## II. Indexing and Querying with Elasticsearch

You should have a better understanding of how to use SQL, now let's leverage the power of Elasticsearch to explore and analyze the data. We'll use the same dataset.


### II.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [5]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below

%pip -q install elasticsearch==7.9.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 213.5/213.5 kB 4.4 MB/s eta 0:00:00


In [6]:
%%bash

wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.0-linux-x86_64.tar.gz
wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.0-linux-x86_64.tar.gz.sha512
tar -xzf elasticsearch-oss-7.9.0-linux-x86_64.tar.gz
sudo chown -R daemon:daemon elasticsearch-7.9.0/
shasum -a 512 -c elasticsearch-oss-7.9.0-linux-x86_64.tar.gz.sha512

elasticsearch-oss-7.9.0-linux-x86_64.tar.gz: OK


In [7]:
%%bash --bg

sudo -H -u daemon elasticsearch-7.9.0/bin/elasticsearch

In [8]:
import time

time.sleep(20)

In [9]:
%%bash

ps -ef | grep elasticsearch

root         919     917  0 15:15 ?        00:00:00 sudo -H -u daemon elasticsearch-7.9.0/bin/elasti
daemon       920     919 37 15:15 ?        00:00:21 /content/elasticsearch-7.9.0/jdk/bin/java -Xshar
root        1362    1360  0 15:16 ?        00:00:00 grep elasticsearch


In [10]:
%%bash

curl -sX GET "http://localhost:9200/_cat/indices?v"

health status index uuid pri rep docs.count docs.deleted store.size pri.store.size


In [11]:
%%bash

curl -sX GET "localhost:9200/"

{
  "name" : "72d0ec9197ac",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "hL8UKgZxTySS2gEpjhZfag",
  "version" : {
    "number" : "7.9.0",
    "build_flavor" : "oss",
    "build_type" : "tar",
    "build_hash" : "a479a2a7fce0389512d6a9361301708b92dff667",
    "build_date" : "2020-08-11T21:36:48.204330Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}


In [12]:
from elasticsearch import Elasticsearch, helpers
import time
import json

In [13]:
es = Elasticsearch(hosts = [{"host":"localhost", "port":9200}])
# Check if python is connected to elasticsearch
es.ping()

True

In [14]:
ES_NODES = "http://localhost:9200"
es = Elasticsearch(hosts = [ES_NODES])

### II.2. Set up Elasticsearch Index

The first step you need to do which is **preprocessing, let's keep it simple and simply drop all row that have missing values.**

Next, you need to set up an Elasticsearch index. An Elasticsearch index is like a database in a traditional relational database. An index lets you store, search, and analyze big volumes of data quickly and in near real time.

You'll also need to setup the mappings for the data types, let's do that here.

In [15]:
df_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [16]:
df_data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [17]:
df_data.dropna(inplace=True)

In [18]:
df_data.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [19]:
# Preprocessing here

# Load your dataset and create an Elasticsearch index
# Hints: use the BulkAPI for faster creation of the index

index_name = 'ecommerce_index'

#### Create Index

- https://opster.com/guides/elasticsearch/glossary/elasticsearch-index/
- https://opster.com/guides/elasticsearch/data-architecture/elasticsearch-create-index-with-mapping/

In [20]:
def create_index(es, index_name):
    # Define mappings for the data types
    mappings = {
          "settings": {
          "analysis": {
            "analyzer": {
              "edge_ngram_analyzer": {
                "tokenizer": "edge_ngram_tokenizer"
              }
            },
            "tokenizer": {
              "edge_ngram_tokenizer": {
                "type": "edge_ngram",
                "min_gram": 2,
                "max_gram": 10,
                "token_chars": [
                  "letter",
                  "digit"
                ]
              }
            }
          }
        },
        "mappings": {
            'properties': {
                'Country': {
                    'type': 'text',
                    'fields': {
                        'keyword': {'type': 'keyword', 'ignore_above': 256}
                    }
                },
                'CustomerID': {
                    'type': 'float'
                },
                'Description': {
                    'type': 'text',
                    'fields': {
                        'keyword': {'type': 'keyword', 'ignore_above': 256}
                    }
                },
                'InvoiceDate': {
                    'type': 'text',
                    'fields': {
                        'keyword': {'type': 'keyword', 'ignore_above': 256}
                    }
                },
                'InvoiceNo': {
                    'type': 'text',
                    'fields': {
                        'keyword': {'type': 'keyword', 'ignore_above': 256}
                    }
                },
                'Quantity': {
                    'type': 'long'
                },
                'StockCode': {
                    'type': 'text',
                    'fields': {
                        'keyword': {'type': 'keyword', 'ignore_above': 256}
                    }
                },
                'UnitPrice': {
                    'type': 'float'
                }
            }
        }
    }

    # Create the index with mappings
    es.indices.create(index=index_name, body=mappings)

In [21]:
create_index(es, index_name)

#### Insert Index

In [22]:
# Insert data

def index_data(es, index_name, data):
    actions = [
        {
            "_index": index_name,
            "_source": {
                "InvoiceNo": row["InvoiceNo"],
                "StockCode": row["StockCode"],
                "Description": row["Description"],
                "Quantity": row["Quantity"],
                "InvoiceDate": row["InvoiceDate"],
                "UnitPrice": row["UnitPrice"],
                "CustomerID": row["CustomerID"],
                "Country": row["Country"]
            }
        }
        for _, row in data.iterrows()
    ]

    # Use the Bulk API for faster indexing
    helpers.bulk(es, actions)

In [23]:
index_data(es, index_name, df_data)

In [24]:
resp = es.indices.get(
    index=index_name,
)
print(resp)

{'ecommerce_index': {'aliases': {}, 'mappings': {'properties': {'Country': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'CustomerID': {'type': 'float'}, 'Description': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'InvoiceDate': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'InvoiceNo': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'Quantity': {'type': 'long'}, 'StockCode': {'type': 'text', 'fields': {'keyword': {'type': 'keyword', 'ignore_above': 256}}}, 'UnitPrice': {'type': 'float'}}}, 'settings': {'index': {'number_of_shards': '1', 'provided_name': 'ecommerce_index', 'creation_date': '1708528638198', 'analysis': {'analyzer': {'edge_ngram_analyzer': {'tokenizer': 'edge_ngram_tokenizer'}}, 'tokenizer': {'edge_ngram_tokenizer': {'token_chars': ['letter', 'digit'], 'min_gram': '2', 'type': 'edge_ngram', 'max_gram': '10'}}}, 'number_of_repl

##### OUTPUT

```Python
{
  'ecommerce_index': {
      'aliases': {},
      'mappings': {
          'properties': {
              'Country': {
                  'type': 'text',
                  'fields': {
                      'keyword': {'type': 'keyword', 'ignore_above': 256}
                  }
              },
              'CustomerID': {
                  'type': 'float'
              },
              'Description': {
                  'type': 'text',
                  'fields': {
                      'keyword': {'type': 'keyword', 'ignore_above': 256}
                  }
              },
              'InvoiceDate': {
                  'type': 'text',
                  'fields': {
                      'keyword': {'type': 'keyword', 'ignore_above': 256}
                  }
              },
              'InvoiceNo': {
                  'type': 'text',
                  'fields': {
                      'keyword': {'type': 'keyword', 'ignore_above': 256}
                  }
              },
              'Quantity': {
                  'type': 'long'
              },
              'StockCode': {
                  'type': 'text',
                  'fields': {
                      'keyword': {'type': 'keyword', 'ignore_above': 256}
                  }
              },
              'UnitPrice': {
                  'type': 'float'
              }
          }
      },
      'settings': {
          'index': {
              'creation_date': '1708223145751',
              'number_of_shards': '1',
              'number_of_replicas': '1',
              'uuid': 'AW0DMe7YRdm9MbYAwToDww',
              'version': {'created': '7090099'},
              'provided_name': 'ecommerce_index'
          }
      }
  }
}

```

field = atribut

1. `ecommerce_index`: Ini adalah nama indeks yang diambil informasinya.
   - `aliases`: Daftar alias yang terkait dengan indeks ini. Alias adalah nama alternatif yang dapat digunakan untuk merujuk ke indeks.
   - `mappings`: Mendefinisikan pemetaan atau struktur dokumen dalam indeks.
     - `properties`: Berisi daftar field-field dalam dokumen dan tipe data dari masing-masing field.
       - `Country, Description, InvoiceDate, InvoiceNo, StockCode`: Field-field dalam dokumen.
         - `type`: Menunjukkan tipe data dari field-field tersebut (misalnya, `text`, `float`, `long`).
         - `fields`: Opsional, biasanya digunakan untuk mendefinisikan sub-field dari tipe `text` (aturan lebih lanjut untuk tipe datanya (type)). Misalnya, `keyword` adalah sub-field dari tipe `text` yang menggunakan tipe data `keyword`.
           - `keyword`: Sub-field yang mungkin digunakan untuk pencarian atau agregasi berbasis keyword.
             - `type`: Tipe data sub-field (misalnya, `keyword`).
             - `ignore_above`: Nilai batas karakter untuk disimpan dalam sub-field `keyword`.
       - `CustomerID, Quantity, UnitPrice`: Field-field lain dalam dokumen dengan tipe data masing-masing.
   - `settings`: Pengaturan konfigurasi untuk indeks.
     - `index`: Konfigurasi khusus untuk indeks ini.
       - `creation_date`: Waktu pembuatan indeks.
       - `number_of_shards`: Jumlah shard yang digunakan dalam indeks.
       - `number_of_replicas`: Jumlah replika (replicas) dari setiap shard dalam indeks.
       - `uuid`: UUID (Universally Unique Identifier) unik yang digunakan untuk mengidentifikasi indeks.
       - `version`: Informasi versi indeks.
         - `created`: Versi Elasticsearch yang digunakan ketika indeks dibuat.
       - `provided_name`: Nama indeks yang disediakan.



- keyword -> ketika search kalimat yang ada di db tidak akan ditokenisasi
- text ->  ketika search kalimat yang ada di db akan ditokenisasi

#### Delete Index

In [ ]:
# response = es.indices.delete(index=index_name, ignore=[400, 404])
# print(response)

{'acknowledged': True}


### II.3. Verifying the index

Before we go to the next section, let's check our index first. You need to change the my_index_name with your chosen index name, afterwards you can just run the code block. The result should be:

- Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
- Number of rows: 406829

In [26]:
from elasticsearch import Elasticsearch

# Connect to the Elasticsearch server
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

index_name = index_name #change the my_index_name value

# Get index mapping (equivalent to getting column names in SQL)
mapping = es.indices.get_mapping(index=index_name)
es_columns = list(mapping[index_name]['mappings']['properties'].keys()) #the variable es_columns is used for submission, do not change the variable name
print(f"Columns: {es_columns}")

# Count documents in the index (equivalent to counting rows in SQL)
es_num_rows = es.count(index=index_name)['count'] #the variable es_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {es_num_rows}")


Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
Number of rows: 406829


### II.4. Running Elasticsearch queries

Once your e-commerce data is cleaned and ready, let's analyse the data to gain insights that will drive the expansion strategy, this time using Elasticsearch and we'll focus on one specific country, which is Germany.

1. **Find all transactions for a specific country**

   Your company operates globally, but you want to understand better how different countries contribute to your sales. By finding all transactions that happened in a specific country, you can get a clearer picture of your company's reach and performance in that location. In our case, we'll focus on the country Germany.


2. **Find all unique products in Germany**

   Your first task is to understand the breadth of your company's product portfolio in Germany. By determining the total number of unique products you sell, you can assess the scale and diversity of your operations.


3. **Find the top 5 most purchased products in Germany**

   Next up, you'll identify the best-selling products in your portfolio in Germany. Knowing which items are the most purchased can help guide decisions about product focus and marketing efforts. We'll limit the result to 5 most purchased products.


By executing these tasks, you will enable your team to provide data-driven insights which will directly contribute to the strategic decisions about your business expansion. As part of the data engineering team, you're not only maintaining the technical infrastructure - you're shaping the future of the company.

There are two more tasks below to check your understanding on how to handle date type data as well as searching using partial keyword:

4. **Find a product's transactions over time**

   Trends in purchases over time for specific products can provide insights into product popularity and seasonality. You decide to track a product's transactions over a period to understand its sales pattern. In this case we'll use the product "I LOVE LONDON MINI RUCKSACK"

5. **Search for a product using a part of its description**

   Finally, you want to investigate how robust and reliable your product search feature is. You decide to test it by searching for a product using a part of its description. In this case, we'll use the keyword "PIN". Return only unique product name. And the product names should be stored in a list in alphabetical order.


- https://gist.github.com/amulyakashyap09/e8442735fe576145dfbed809bb089056

- https://www.elastic.co/search-labs/blog/articles/lexical-and-semantic-search-with-elasticsearch

In [25]:
# Write your Elasticsearch Query here

#use the following variable name to store the result from your Elasticsearch Query
es_total_unique_products = 0
es_transactions_in_germany = 0
es_top_products = {} #dictionary where key is product name, value is the quantity sold, see example below
es_transactions_over_time = {} #dictionary where key is invoice date, value is the quantity sold, see example below
es_unique_product_search = [] #list containing the products in alphabetical order

#### Nomor 1

In [27]:
# 1. Find all transactions for a specific country (Germany)
# Example: es_transactions_in_germany = 7584

query = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    }
}

search_results = es.search(index=index_name, body=query)

print(json.dumps(search_results))
print(json.dumps(search_results['hits']['total']['value']))

transactions = search_results['hits']['hits']
for transaction in transactions:
    print(transaction['_source']['Description'])

es_transactions_in_germany = search_results['hits']['total']['value']

{"took": 41, "timed_out": false, "_shards": {"total": 1, "successful": 1, "skipped": 0, "failed": 0}, "hits": {"total": {"value": 9495, "relation": "eq"}, "max_score": 4.655447, "hits": [{"_index": "ecommerce_index", "_type": "_doc", "_id": "CeI-zI0BiBXz7Q6AtADQ", "_score": 4.655447, "_source": {"InvoiceNo": "536527", "StockCode": "22809", "Description": "SET OF 6 T-LIGHTS SANTA", "Quantity": 6, "InvoiceDate": "12/1/2010 13:04", "UnitPrice": 2.95, "CustomerID": 12662.0, "Country": "Germany"}}, {"_index": "ecommerce_index", "_type": "_doc", "_id": "CuI-zI0BiBXz7Q6AtADQ", "_score": 4.655447, "_source": {"InvoiceNo": "536527", "StockCode": "84347", "Description": "ROTATING SILVER ANGELS T-LIGHT HLDR", "Quantity": 6, "InvoiceDate": "12/1/2010 13:04", "UnitPrice": 2.55, "CustomerID": 12662.0, "Country": "Germany"}}, {"_index": "ecommerce_index", "_type": "_doc", "_id": "C-I-zI0BiBXz7Q6AtADQ", "_score": 4.655447, "_source": {"InvoiceNo": "536527", "StockCode": "84945", "Description": "MULTI 

##### OUTPUT


```Python
{
    "took": 9,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 9495,
            "relation": "eq"
        },
        "max_score": 4.655447,
        "hits": [{
            "_index": "ecommerce_index",
            "_type": "_doc",
            "_id": "tfEIuo0BhmGmg7Ax2AVI",
            "_score": 4.655447,
            "_source": {...}
        }]
    }
}
```

1. `took`: Menunjukkan berapa lama waktu yang dibutuhkan oleh Elasticsearch untuk mengeksekusi query. Dalam kasus ini, waktu eksekusi adalah 9 milidetik.

2. `timed_out`: Menunjukkan apakah eksekusi query telah melampaui waktu maksimum yang ditetapkan. Jika nilainya adalah `false`, itu berarti query telah selesai dalam waktu yang ditetapkan.

3. `_shards`: Berisi informasi tentang pemecahan (shards) indeks Elasticsearch.
   - `total`: Jumlah total shard yang terlibat dalam operasi.
   - `successful`: Jumlah shard yang berhasil dalam operasi.
   - `skipped`: Jumlah shard yang dilewati dalam operasi.
   - `failed`: Jumlah shard yang gagal dalam operasi.

4. `hits`: Berisi detail tentang dokumen-dokumen yang cocok dengan query.
   - `total`: Informasi tentang jumlah total dokumen yang cocok dengan query.
     - `value`: Jumlah total dokumen yang cocok.
     - `relation`: Menunjukkan hubungan antara nilai `value` dengan jumlah total dokumen yang cocok. Nilai "eq" menunjukkan bahwa jumlah total dokumen yang cocok sama dengan nilai `value`.
   - `max_score`: Nilai skor tertinggi dari dokumen yang cocok.
   - `hits`: Array yang berisi dokumen-dokumen yang cocok dengan query.
     - `_index`: Nama indeks tempat dokumen disimpan.
     - `_type`: Tipe dokumen (deprecated, sebaiknya gunakan `_doc`).
     - `_id`: ID unik dokumen.
     - `_score`: Skor relevansi dokumen terhadap query.
     - `_source`: Sumber dokumen yang berisi field dan nilai-fieldnya.



#### Nomor 2

- https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html
- https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html
- https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html

- https://opster.com/guides/elasticsearch/search-apis/elasticsearch-search-unique-values-aggregations-cardinality/
- https://opster.com/guides/elasticsearch/search-apis/elasticsearch-count-distinct/


In [29]:
# 2. Find all unique products in Germany
# Hints: Use set() to remove duplicate values
# Example: es_total_unique_products = 45345

query = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    },
    "aggs": {
        "total_unique_products": {
            "cardinality": {
                "field": "Description.keyword",
            }
        }
    }
}

search_results = es.search(index=index_name, body=query)

print(search_results)
print(search_results['aggregations']['total_unique_products']['value'])

es_total_unique_products = search_results['aggregations']['total_unique_products']['value']

{'took': 62, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 9495, 'relation': 'eq'}, 'max_score': 4.655447, 'hits': [{'_index': 'ecommerce_index', '_type': '_doc', '_id': 'q1k0zI0Bv-3rpF2k5nwy', '_score': 4.655447, '_source': {'InvoiceNo': 'C547581', 'StockCode': '22367', 'Description': 'CHILDRENS APRON SPACEBOY DESIGN', 'Quantity': -1, 'InvoiceDate': '3/24/2011 10:58', 'UnitPrice': 1.95, 'CustomerID': 12474.0, 'Country': 'Germany'}}, {'_index': 'ecommerce_index', '_type': '_doc', '_id': 'rFk0zI0Bv-3rpF2k5nwy', '_score': 4.655447, '_source': {'InvoiceNo': 'C547581', 'StockCode': '22554', 'Description': 'PLASTERS IN TIN WOODLAND ANIMALS', 'Quantity': -1, 'InvoiceDate': '3/24/2011 10:58', 'UnitPrice': 1.65, 'CustomerID': 12474.0, 'Country': 'Germany'}}, {'_index': 'ecommerce_index', '_type': '_doc', '_id': 'rVk0zI0Bv-3rpF2k5nwy', '_score': 4.655447, '_source': {'InvoiceNo': 'C547581', 'StockCode': '21561', 'Description

In [30]:
coba = df_data[df_data['Country'] == 'Germany']

coba['Description'].nunique()

1703

##### OUTPUT

```Python
{
    'took': 33,
    'timed_out': False,
    '_shards': {
        'total': 1,
        'successful': 1,
        'skipped': 0,
        'failed': 0
    },
    'hits': {
        'total': {
            'value': 9495,
            'relation': 'eq'
        },
        'max_score': 4.655447,
        'hits': [{
            '_index': 'ecommerce_index',
            '_type': '_doc',
            '_id': 'tfEIuo0BhmGmg7Ax2AVI',
            '_score': 4.655447,
            '_source': {...}
        }]
    },
    'aggregations': {
        'total_unique_products': {
            'value': 1703
        }
    }
}
```



#### Nomor 3

- https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html

In [31]:
# 3. Find the top 5 most purchased products in Germany
# Example: es_top_products = {'PRODUCT 1': 498, 'PRODUCT 2': 452, 'PRODUCT 3': 342, 'PRODUCT 4': 231, 'PRODUCT 5': 123}

query = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    },
    "aggs": {
        "top_products": {
            "terms": {
                "field": "Description.keyword",  # Assuming 'StockCode' represents products
                "size": 5,  # Limit to top 5 products
                "order": {"_count": "desc"}  # Order by quantity sold in descending order
            }
        }
    }
}

search_results = es.search(index=index_name, body=query)

print(search_results)

top_products = {}
for bucket in search_results['aggregations']['top_products']['buckets']:
    top_products[bucket['key']] = bucket['doc_count']

print(top_products)
es_top_products = top_products

# Print the top 5 most purchased products in Germany
print("Top 5 most purchased products in Germany:")
for product, quantity in top_products.items():
    print(f"{product}: {quantity}")

{'took': 46, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 9495, 'relation': 'eq'}, 'max_score': 4.655447, 'hits': [{'_index': 'ecommerce_index', '_type': '_doc', '_id': 'q1k0zI0Bv-3rpF2k5nwy', '_score': 4.655447, '_source': {'InvoiceNo': 'C547581', 'StockCode': '22367', 'Description': 'CHILDRENS APRON SPACEBOY DESIGN', 'Quantity': -1, 'InvoiceDate': '3/24/2011 10:58', 'UnitPrice': 1.95, 'CustomerID': 12474.0, 'Country': 'Germany'}}, {'_index': 'ecommerce_index', '_type': '_doc', '_id': 'rFk0zI0Bv-3rpF2k5nwy', '_score': 4.655447, '_source': {'InvoiceNo': 'C547581', 'StockCode': '22554', 'Description': 'PLASTERS IN TIN WOODLAND ANIMALS', 'Quantity': -1, 'InvoiceDate': '3/24/2011 10:58', 'UnitPrice': 1.65, 'CustomerID': 12474.0, 'Country': 'Germany'}}, {'_index': 'ecommerce_index', '_type': '_doc', '_id': 'rVk0zI0Bv-3rpF2k5nwy', '_score': 4.655447, '_source': {'InvoiceNo': 'C547581', 'StockCode': '21561', 'Description

##### OUTPUT

```Python
{
    'took': 33,
    'timed_out': False,
    '_shards': {
        'total': 1,
        'successful': 1,
        'skipped': 0,
        'failed': 0
    },
    'hits': {
        'total': {
            'value': 9495,
            'relation': 'eq'
        },
        'max_score': 4.655447,
        'hits': [{
            '_index': 'ecommerce_index',
            '_type': '_doc',
            '_id': 'tfEIuo0BhmGmg7Ax2AVI',
            '_score': 4.655447,
            '_source': {...}
        }]
    },
    'aggregations': {
        'top_products': {
            'doc_count_error_upper_bound': 0,
            'sum_other_doc_count': 8766,
            'buckets': [
              {'key': 'POSTAGE', 'doc_count': 383},
              {'key': 'ROUND SNACK BOXES SET OF4 WOODLAND ', 'doc_count': 120},
              {'key': 'REGENCY CAKESTAND 3 TIER', 'doc_count': 81},
              {'key': 'ROUND SNACK BOXES SET OF 4 FRUITS ', 'doc_count': 78},
              {'key': 'PLASTERS IN TIN WOODLAND ANIMALS', 'doc_count': 67}
            ]
        }
    }
}
```



#### Nomor 4

- https://opster.com/guides/elasticsearch/search-apis/elasticsearch-match-multi-match-and-match-phrase-queries/

In [32]:
# 4. Find a product's transactions over time
# Hints: Date format is M/d/yyyy H:m
# Product = I LOVE LONDON MINI RUCKSACK
# Example: es_transactions_over_time = {'12/12/2010 11:11': 1, 12/12/2010 11:12': 1}
#
product_name = "I LOVE LONDON MINI RUCKSACK"

# query = {
#     "query": {
#         "match": {
#             "Description": product_name
#         }
#     }
#

# query = {
#      "query": {
#          "bool": {
#              "must": [{
#                  "match": {
#                      "Description": product_name
#                  }
#              }]
#          },
#          "wildcard": {
#             "Description": {
#                 "value": f"{product_name}"
#             }
#         }
#     }
# }

query = {
    "query": {
        "match_phrase_prefix": { #fuzzy based (token)
            "Description": {
                "query": "I LOVE LONDON MINI RUCKSACK"
            }
        }
    }
}

search_results = es.search(index=index_name, body=query)

print(search_results['hits']['hits'])

transactions_over_time = {}
for hit in search_results['hits']['hits']:
    timestamp = hit['_source']['InvoiceDate']
    transactions_over_time[timestamp] = transactions_over_time.get(timestamp, 0) + 1


print(transactions_over_time)
es_transactions_over_time = transactions_over_time

# print(f"Transactions over time for product '{product_name}':")
for timestamp, count in transactions_over_time.items():
    print(f"{timestamp}: {count}")

[{'_index': 'ecommerce_index', '_type': '_doc', '_id': '71w1zI0Bv-3rpF2kKmdv', '_score': 27.338432, '_source': {'InvoiceNo': '569720', 'StockCode': '23391', 'Description': ' I LOVE LONDON MINI RUCKSACK', 'Quantity': 1, 'InvoiceDate': '10/6/2011 9:11', 'UnitPrice': 4.15, 'CustomerID': 14646.0, 'Country': 'Netherlands'}}]
{'10/6/2011 9:11': 1}
10/6/2011 9:11: 1


#### Nomor 5

- https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html
- https://opster.com/guides/elasticsearch/search-apis/elasticsearch-strings-keyword-vs-text-vs-wildcard/

- https://stackoverflow.com/questions/58145815/keyword-searching-and-filtering-in-elasticsearch

- https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-wildcard-query.html

- https://opster.com/guides/elasticsearch/search-apis/elasticsearch-wildcard-queries/

- https://stackoverflow.com/questions/6467067/how-to-search-for-a-part-of-a-word-with-elasticsearch

In [33]:
# 5. Search for a product using a part of its description
# Hints: Return only unique Product names, see number 1
# Example: es_unique_product_search = ['PRODUCT 1', 'PRODUCT 2']

keyword = "PIN"

query = {
    "query": {
        "wildcard": {
            "Description": {
                "value": f"{keyword}"
                # "value": f"*{keyword}*"
            }
        }
    }
}

search_results_ = es.search(index=index_name, body=query)
print(search_results_)

unique_product_names_ = set()
for hit in search_results_['hits']['hits']:
    print(hit)
    unique_product_names_.add(hit['_source']['Description'])

unique_product_names_ = sorted(list(unique_product_names_))

print(unique_product_names_)

# query = {
#   "query": {
#     "fuzzy": {
#       "Description": {
#         "value": "PIN",
#         "fuzziness": "AUTO"
#       }
#     }
#   }
# }

# query = {
#   "query": {
#     "match": {
#       "Description": "PIN"
#     }
#   }
# }

# query = {
#   "query": {
#     "match_phrase": {
#       "Description": "PIN"
#     }
#   }
# }

# query = {
#   "query": {
#     "wildcard": {
#       "Description": "PIN"
#     }
#   }
# }

query = {
  "query": {
    "match": {
      "Description": "PIN"
    }
  }
}

search_results = es.search(index=index_name, body=query)

print(search_results)

unique_product_names = set()
for hit in search_results['hits']['hits']:
    print(hit)
    unique_product_names.add(hit['_source']['Description'])

unique_product_names = sorted(list(unique_product_names))

print(unique_product_names)
# es_unique_product_search = unique_product_names

print("Unique product names containing 'PIN' in the description:")
for product_name in unique_product_names:
    print(product_name)


es_unique_product_search = unique_product_names + unique_product_names_
print(es_unique_product_search)
es_unique_product_search.sort()
print(es_unique_product_search)

{'took': 5, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 355, 'relation': 'eq'}, 'max_score': 1.0, 'hits': [{'_index': 'ecommerce_index', '_type': '_doc', '_id': '3uI-zI0BiBXz7Q6AtgQW', '_score': 1.0, '_source': {'InvoiceNo': '536633', 'StockCode': '22446', 'Description': 'PIN CUSHION BABUSHKA PINK', 'Quantity': 3, 'InvoiceDate': '12/2/2010 11:20', 'UnitPrice': 3.35, 'CustomerID': 14901.0, 'Country': 'United Kingdom'}}, {'_index': 'ecommerce_index', '_type': '_doc', '_id': 'nOI-zI0BiBXz7Q6Avig1', '_score': 1.0, '_source': {'InvoiceNo': '537765', 'StockCode': '22448', 'Description': 'PIN CUSHION BABUSHKA RED', 'Quantity': 1, 'InvoiceDate': '12/8/2010 12:08', 'UnitPrice': 3.35, 'CustomerID': 14606.0, 'Country': 'United Kingdom'}}, {'_index': 'ecommerce_index', '_type': '_doc', '_id': 'BuI-zI0BiBXz7Q6Ay3qf', '_score': 1.0, '_source': {'InvoiceNo': '540550', 'StockCode': '22448', 'Description': 'PIN CUSHION BABUSHKA RE

In [ ]:
# es.close()

## Submission

Once you are satisfied with the performance of your model, then you run the code block below to submit your project.


In [41]:
# Submit Method

# Do not change the code below
# question_id = "01_sql_columns"
# submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
# question_id = "02_sql_num_rows"
# submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
# question_id = "03_sql_num_products"
# submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
# question_id = "04_sql_num_customers"
# submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
# question_id = "05_sql_top_product_revenues"
# submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
# question_id = "06_sql_top_countries"
# submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
# question_id = "07_sql_top_selling_products_by_country_by_sales"
# submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
# question_id = "08_sql_top_selling_products_by_country_by_quantity"
# submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)

# question_id = "09_es_columns"
# submit(student_id, name, assignment_id, str(es_columns), question_id, drive_link)
# question_id = "10_es_num_rows"
# submit(student_id, name, assignment_id, str(es_num_rows), question_id, drive_link)
# question_id = "11_es_total_unique_products"
# submit(student_id, name, assignment_id, str(es_total_unique_products), question_id, drive_link)
# question_id = "12_es_transactions_in_germany"
# submit(student_id, name, assignment_id, str(es_transactions_in_germany), question_id, drive_link)
# question_id = "13_es_top_products"
# submit(student_id, name, assignment_id, str(es_top_products), question_id, drive_link)
# question_id = "14_es_transactions_over_time"
# submit(student_id, name, assignment_id, str(es_transactions_over_time), question_id, drive_link)
question_id = "15_es_unique_product_search"
submit(student_id, name, assignment_id, str(es_unique_product_search), question_id, drive_link)

'Assignment successfully submitted'

## FIN